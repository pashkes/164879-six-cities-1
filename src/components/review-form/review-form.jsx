import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {compose} from "recompose";

import Rating from "../rating/rating.jsx";
import withReviewForm from "./../../hocs/with-review-form/with-review-form";

import {Operation, ActionCreators} from "../../reducer/data/data";

import {
  getStatusSendingReview,
  getStatusIsSentReview,
  getError,
} from "./../../reducer/data/selectors";
import {KeyCode} from "../../constants";
import {CommentLength} from "../../constants";

export class ReviewForm extends PureComponent {
  constructor(props) {
    super(props);
    this.handleSubmitForm = this.handleSubmitForm.bind(this);
    this.handleMessage = this.handleMessage.bind(this);
    this.handleMessageKeyDown = this.handleMessageKeyDown.bind(this);
    this.handleMessageOutside = this.handleMessageOutside.bind(this);
    this.form = React.createRef();
  }

  handleSubmitForm(evt) {
    evt.preventDefault();
    const {sendComment, rating, comment} = this.props;
    sendComment({rating, comment});
  }

  handleMessage() {
    document.addEventListener(`keydown`, this.handleMessageKeyDown);
  }

  handleMessageKeyDown({ctrlKey, keyCode}) {
    const {
      sendComment,
      rating,
      comment,
      isFormValid,
    } = this.props;
    if (isFormValid && ctrlKey && keyCode === KeyCode.ENTER) {
      sendComment({rating, comment});
    }
  }

  handleMessageOutside() {
    document.removeEventListener(`keydown`, this.handleMessageKeyDown);
  }

  componentDidUpdate() {
    const {isReviewSent, updateForm} = this.props;
    if (isReviewSent) {
      updateForm();
      this.form.current.reset();
    }
  }

  render() {
    const {
      onChangeMessage,
      onChangeRating,
      error,
      submitButton,
      textarea,
      isReviewSending,
      isFormValid,
    } = this.props;
    return (
      <form
        className="reviews__form form"
        onSubmit={this.handleSubmitForm}
        action="#"
        method="post"
        ref={this.form}
      >
        <label className="reviews__label form__label" htmlFor="review">Your review</label>
        <Rating onChangeRating={onChangeRating}/>
        <textarea
          ref={textarea}
          onChange={onChangeMessage}
          onFocus={this.handleMessage}
          onBlur={this.handleMessageOutside}
          minLength={CommentLength.MIN}
          maxLength={CommentLength.MAX}
          className="reviews__textarea form__textarea"
          id="review"
          name="review"
          placeholder="Tell how was your stay, what you like and what can be improved"/>
        <span role="alert" aria-live="assertive">{error ? error : ``}</span>
        <div className="reviews__button-wrapper">
          <p className="reviews__help">
            To submit review please make sure to set <span className="reviews__star">rating</span> and
            describe your stay with at least <b className="reviews__text-amount">{CommentLength.MIN} characters</b>.
          </p>
          <button
            ref={submitButton}
            className="reviews__submit form__submit button"
            type="submit"
            disabled={!(isFormValid && !isReviewSending)}
          >{isReviewSending ? `Sending...` : `Submit`}
          </button>
        </div>
      </form>
    );
  }
}

ReviewForm.propTypes = {
  onChangeMessage: PropTypes.func.isRequired,
  onChangeRating: PropTypes.func.isRequired,
  idCurrentOffer: PropTypes.number.isRequired,
  rating: PropTypes.number.isRequired,
  comment: PropTypes.string.isRequired,
  sendComment: PropTypes.func.isRequired,
  submitButton: PropTypes.object.isRequired,
  textarea: PropTypes.object.isRequired,
  updateForm: PropTypes.func.isRequired,
  isReviewSending: PropTypes.bool.isRequired,
  isReviewSent: PropTypes.bool.isRequired,
  error: PropTypes.string,
  isFormValid: PropTypes.bool.isRequired,
};

const mapStateToProps = (state) => ({
  isReviewSending: getStatusSendingReview(state),
  isReviewSent: getStatusIsSentReview(state),
  error: getError(state),
});

const mapDispatchToProps = (dispatch, ownProps) => ({
  sendComment: (data)=> {
    dispatch(Operation.postReview(ownProps.idCurrentOffer, data));
    dispatch(ActionCreators.lockForm(true));
  },
  updateForm: () => dispatch(ActionCreators.cleanForm(false)),
});

const reviewForm = compose(
    connect(mapStateToProps, mapDispatchToProps),
    withReviewForm,
);
export default reviewForm(ReviewForm);
