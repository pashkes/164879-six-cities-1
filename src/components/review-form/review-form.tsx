import * as React from "react";
import {connect} from "react-redux";
import {compose} from "recompose";

import withReviewForm from "./../../hocs/with-review-form/with-review-form";
import {Operation, ActionCreators} from "../../reducer/data/data";
import {
  getStatusSendingReview,
  getStatusIsSentReview,
  getError,
} from "./../../reducer/data/selectors";
import {KeyCode} from "../../constants";
import {CommentLength} from "../../constants";

import Rating from "../rating/rating";

interface Props {
  onChangeMessage: () => void,
  onChangeRating: () => void,
  idCurrentOffer: number,
  rating: number,
  comment: string,
  onSendComment: ({rating: number, comment: string}) => void,
  onUpdateForm: () => void,
  isReviewSending: boolean,
  isReviewSent: boolean,
  error: string,
  isFormValid: boolean,
}

export class ReviewForm extends React.PureComponent<Props> {
  readonly _form: React.RefObject<HTMLFormElement>;
  constructor(props) {
    super(props);

    this._form = React.createRef();
  }

  handleSubmitForm = (evt) => {
    evt.preventDefault();
    const {onSendComment, rating, comment} = this.props;

    onSendComment({rating, comment});
  };

  handleMessageKeyDown = ({ctrlKey, keyCode}) => {
    const {
      onSendComment,
      rating,
      comment,
      isFormValid,
    } = this.props;

    if (isFormValid && ctrlKey && keyCode === KeyCode.ENTER) {
      onSendComment({rating, comment});
    }
  };

  componentDidUpdate() {
    const {isReviewSent, onUpdateForm} = this.props;

    if (isReviewSent) {
      onUpdateForm();
      this._form.current.reset();
    }
  }

  render() {
    const {
      onChangeMessage,
      onChangeRating,
      error,
      isReviewSending,
      isFormValid,
      comment,
    } = this.props;

    return (
      <form
        className="reviews__form form"
        onSubmit={this.handleSubmitForm}
        action="#"
        method="post"
        ref={this._form}
      >
        <label className="reviews__label form__label" htmlFor="review">Your review</label>
        <Rating onChangeRating={onChangeRating}/>
        <textarea
          value={comment}
          onChange={onChangeMessage}
          onKeyDown={this.handleMessageKeyDown}
          minLength={CommentLength.MIN}
          maxLength={CommentLength.MAX}
          className="reviews__textarea form__textarea"
          id="review"
          name="review"
          placeholder="Tell how was your stay, what you like and what can be improved"/>
        <span role="alert" aria-live="assertive">{error ? error : ``}</span>
        <div className="reviews__button-wrapper">
          <p className="reviews__help">
            To submit review please make sure to set <span className="reviews__star">rating</span> and
            describe your stay with at least <b className="reviews__text-amount">{CommentLength.MIN} characters</b>.
          </p>
          <button
            className="reviews__submit form__submit button"
            type="submit"
            disabled={!(isFormValid && !isReviewSending)}
          >{isReviewSending ? `Sending...` : `Submit`}
          </button>
        </div>
      </form>
    );
  }
}

const mapStateToProps = (state) => ({
  isReviewSending: getStatusSendingReview(state),
  isReviewSent: getStatusIsSentReview(state),
  error: getError(state),
});

const mapDispatchToProps = (dispatch, ownProps) => ({
  onSendComment: (data) => {
    dispatch(Operation.postReview(ownProps.idCurrentOffer, data));
    dispatch(ActionCreators.lockForm(true));
  },
  onUpdateForm: () => dispatch(ActionCreators.cleanForm(false)),
});

const enhance = compose(
    connect(mapStateToProps, mapDispatchToProps),
    withReviewForm,
);
export default enhance(ReviewForm);
