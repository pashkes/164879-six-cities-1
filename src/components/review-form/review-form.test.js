import React from "react";
import {shallow, configure} from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import {ReviewForm} from './review-form.jsx';
configure({adapter: new Adapter()});

it(`Review form renders correctly`, () => {
  const reviewForm = shallow(<ReviewForm
    onChangeMessage={jest.fn()}
    onChangeRating={jest.fn()}
    idCurrentOffer={0}
    rating={0}
    comment={``}
    onSendComment={jest.fn()}
    onUpdateForm={jest.fn()}
    isReviewSending={false}
    isReviewSent={false}
    isFormValid={true}
    error={``}
  />);
  expect(reviewForm).toMatchSnapshot();
});
