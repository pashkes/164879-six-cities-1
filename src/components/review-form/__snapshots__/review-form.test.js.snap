// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review form renders correctly 1`] = `
<form
  action="#"
  className="reviews__form form"
  method="post"
  onSubmit={[Function]}
>
  <label
    className="reviews__label form__label"
    htmlFor="review"
  >
    Your review
  </label>
  <Rating
    onChangeRating={[MockFunction]}
  />
  <textarea
    className="reviews__textarea form__textarea"
    id="review"
    maxLength={300}
    minLength={50}
    name="review"
    onBlur={[Function]}
    onChange={[MockFunction]}
    onFocus={[Function]}
    placeholder="Tell how was your stay, what you like and what can be improved"
  />
  <span
    aria-live="assertive"
    role="alert"
  />
  <div
    className="reviews__button-wrapper"
  >
    <p
      className="reviews__help"
    >
      To submit review please make sure to set 
      <span
        className="reviews__star"
      >
        rating
      </span>
       and describe your stay with at least 
      <b
        className="reviews__text-amount"
      >
        50
         characters
      </b>
      .
    </p>
    <button
      className="reviews__submit form__submit button"
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
